ARCs/arc-0056.md

```diff
@@ -0,0 +1,343 @@
+
+```ts
+/** Describes a method in the contract. This interface is an extension of the interface described in ARC-4 */
+interface Method {
+  /** The name of the method */
+  name: string;
+  /** Optional, user-friendly description for the method */
+  desc?: string;
+  /** The arguments of the method, in order */
+  args: Array<{
+    /** The type of the argument */
+    type: ABIType;
+    /** If the type is a struct, the name of the struct */
+    struct?: StructName;
+    /** Optional, user-friendly name for the argument */
+    name?: string;
+    /** Optional, user-friendly description for the argument */
+    desc?: string;
+    /** The default value that clients should use. MUST be base64 encoded bytes */
+    defaultValue?: string;
+  }>;
+  /** Information about the method's return value */
+  returns: {
+    /** The type of the return value, or "void" to indicate no return value. */
+    type: ABIType;
+    /** If the type is a struct, the name of the struct */
+    struct?: StructName;
+    /** Optional, user-friendly description for the return value */
+    desc?: string;
+  };
+  /** an action is a combination of call/create and an OnComplete */
+  actions: {
+    /** OnCompletes this method allows when appID === 0 */
+    create: ("NoOp" | "OptIn" | "DeleteApplication")[];
+    /** OnCompletes this method allows when appID !== 0 */
+    call: (
+      | "NoOp"
+      | "OptIn"
+      | "CloseOut"
+      | "ClearState"
+      | "UpdateApplication"
+      | "DeleteApplication"
+    )[];
+  };
+  /** If this method does not write anything to the ledger (ARC-22) */
+  readonly?: boolean;
+  /** ARC-28 events that MAY be emitted by this method */
+  events?: Array<Event>;
+  /** Information that clients can use when calling the method */
+  recommendations?: {
+    /** The number of inner transactions the caller should cover the fees for */
+    innerTransactionCount?: number;
+    /** Recommended box references to include */
+    boxes?: {
+      /** The app ID for the box */
+      app?: number;
+      /** The base64 encoded box key */
+      key: string;
+      /** The number of bytes being read from the box */
+      readBytes: number;
+      /** The number of bytes being written to the box */
+      writeBytes: number;
+    };
+    /** Recommended foreign accounts */
+    accounts?: string[];
+    /** Recommended foreign apps */
+    apps?: number[];
+    /** Recommended foreign assets */
+    assets?: number[];
+  };
+}
+```
+
+### Event Interface
+
+[ARC-28](./arc-0028.md) events are described using an extension of the original interface described in the ARC, with the addition of an optional struct field for arguments
+
+```ts
+interface Event {
+  /** The name of the event */
+  name: string;
+  /** Optional, user-friendly description for the event */
+  desc?: string;
+  /** The arguments of the event, in order */
+  args: Array<{
+    /** The type of the argument */
+    type: ABIType;
```

Comment:
Since this isn't part of ARC-4, maybe in the case of a struct, we can avoid re-specifying it as a tuple here?

Reply:
It is not a part of ARC4 but it is a part of ARC28. You could make the argument that ARC28 doesn't have much adoption yet, but I think it's better to keep backwards compatibility with existing ARCs